tools/cloud-function-automation/cloud_function_automation.py

```diff
@@ -0,0 +1,278 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Usage: python3 cloud_function_automation.py
+
+This script will be used to workflow/pipeline where user can
+route filtered logs to pubsub and trigger cloud function
+based on pub/sub events.
+
+resource => log sink => pub/sub => cloud function
+
+Please update all config variables in variables.py file.
+"""
+
+import time
+import json
+from getpass import getpass
+from google.api_core.exceptions import AlreadyExists
+import requests
+from google.cloud import pubsub_v1, logging, functions_v1
+from variables import PROJECT_ID, PUBSUB_TOPIC_NAME, SINK_NAME, \
+    LOG_SINK_FILTER, CLOUD_FUNCTION_CONFIG, PROJECT_NUMBER, ACTION
+
+
+def enable_required_service_apis(project_num, token):
+    """
+    This function will be used to enable all services required in this script.
+    :param project_num: Google Cloud project number mentioned in variables.py.
+    :param token: Google Cloud auth token.
+    :return: True or False.
+    """
+    url = f"https://serviceusage.googleapis.com/v1/projects/{project_num}/services:batchEnable"
+
+    payload = json.dumps({
+        "serviceIds": [
+            "pubsub.googleapis.com",
+            "cloudfunctions.googleapis.com",
+            "logging.googleapis.com"
+        ]
+    })
+    headers = {
+        'Authorization': token,
+        'Content-Type': 'application/json'
+    }
+    try:
+        response = requests.request("POST", url, headers=headers, data=payload)
+    except requests.exceptions.Timeout as exc:
+        raise SystemExit(f"Timeout Error while calling enable service API - {exc}")
+    except requests.exceptions.TooManyRedirects as exc:
+        raise SystemExit(f"TooManyRedirects Error while calling enable service API - {exc}")
+    except requests.exceptions.RequestException as exc:
+        raise SystemExit(f"RequestException while calling enable service API - {exc}")
+
+    if response.status_code == 200:
+        services_enabled = True
+    else:
+        services_enabled = False
+
+    return services_enabled
+
+
+def create_pubsub_topic(project_id, pubsub_topic_name):
+    """
+    This function will be used to create pubsub topic if not already present.
+    :param project_id: Google Cloud project id mentioned in variables.py.
+    :param pubsub_topic_name: Google Cloud Pub/Sub topic name mentioned in variables.py.
+    :return: True or False.
+    """
+    try:
+        publisher = pubsub_v1.PublisherClient()
+    except Exception as exc:
```

Comment:
What kind of exception can happen here? Handling a generic exception and then exiting immediately does not do much. Please handle the exceptions that you would know how to handle, or at least, that you could enrich with some more meaningful error message.

Reply:
removed all generic exception handling